ci: Move Windows MSVC build from AppVeyor to Cirrus

hebasto opened the pull request:
This PR:
- gets rid off unreliable AppVeyor CI
- places all CI configs in one place
- allows to enable functional tests in the future (using a persistent worker)

| no populated `vcpkg` cache | populated `vcpkg` cache |
|---|---|
| ![Screenshot from 2021-09-02 15-47-04](https://user-images.githubusercontent.com/32963518/131846156-9367bffc-9093-40ca-98c3-15db74e24113.png) | ![Screenshot from 2021-09-02 14-06-26](https://user-images.githubusercontent.com/32963518/131833053-a501454d-eecf-463c-a3a4-b89d2a494058.png) |


Currently, AppVeyor builds take about 44..48 minutes.

hebasto commented:
The warmed up vcpkg cache [reduces](https://cirrus-ci.com/task/6335159239704576) build time by about 40 minutes.

MarcoFalke commented:
Still takes longer than the appveyor 38 minutes right now?

hebasto commented:
> Still takes longer than the appveyor 38 minutes right now?

Right.

hebasto commented:
@MarcoFalke 
> Still takes longer than the appveyor 38 minutes right now?

Does it make sense to move this task into the DrahtBot infrastructure? Or just close it?

hebasto commented:
> Still takes longer than the appveyor 38 minutes right now?

It takes 27 minutes now.

> The time is spent on installing `msbuild` tools, that are preinstalled on AppVeyor.

Fixed by using `visualstudio2019` image.

fanquake commented:
What is the status of this? Is the intention to replace Appveyor? If not, do we need two Windows builds? 

An empty PR description is not really useful.

MarcoFalke commented:
(Also needs appveyor removed)

hebasto commented:
@sipsorcery 

It seems [`Qt5.12.11_x64_static_vs2019_16101.zip`](https://github.com/sipsorcery/qt_win_binary/releases/download/qt51211x64_static_vs2019_16101/Qt5.12.11_x64_static_vs2019_16101.zip) is not compatible with the `msbuild` in the `cirrusci/windowsservercore:visualstudio2019` image:

```
LINK : fatal error C1900: Il mismatch between 'P1' version '20210202' and 'P2' version '20210113' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
LINK : fatal error LNK1257: code generation failed [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
LINK : fatal error LNK1327: failure during running link.exe [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
```

So, I switched back to [`Qt5.12.10_x64_static_vs2019_1694.zip`](https://github.com/sipsorcery/qt_win_binary/releases/download/qt51210x64_vs2019_1694/Qt5.12.10_x64_static_vs2019_1694.zip) for now.

May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?

sipsorcery commented:
> May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?

Sure, any idea where I can find the software versions installed on the cirrusci image you're using?

hebasto commented:
> > May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?
> 
> Sure, any idea where I can find the software versions installed on the cirrusci image you're using?

https://github.com/cirruslabs/docker-images-windows

hebasto commented:
@sipsorcery 

The `msvc -version` output within the `cirrusci/windowsservercore:visualstudio2019` Cirrus image:

```
Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
16.9.0.16703

sipsorcery commented:
> The `msvc -version` output within the `cirrusci/windowsservercore:visualstudio2019` Cirrus image:
> 
> ```
> Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET Framework
> Copyright (C) Microsoft Corporation. All rights reserved.
> 16.9.0.16703
> ```

Thanks for doing that. I was trying to track it down via the docker files. I already have that version of msbuild installed so will attempt the Qt build tonight (it kills my machine so overnight job).

hebasto commented:
@sipsorcery 

To make debugging a bit easier in the future, I've added
```
  build_environment_script:
    - choco list --localonly
    - msbuild -version
```

sipsorcery commented:
@hebasto https://github.com/sipsorcery/qt_win_binary/releases/tag/qt51211x64_static_vs2019_160900

hebasto commented:
> @hebasto https://github.com/sipsorcery/qt_win_binary/releases/tag/qt51211x64_static_vs2019_160900

Thanks!

hebasto commented:
@MarcoFalke @fanquake @sipsorcery 

This PR is ready for reviewing now :tiger2:

sipsorcery commented:
ACK ca9b9b9a2c0dd7ab767788be6512cfc34f1f6ee4.

Makes sense to consolidate CI platforms.

Tested that latest msbuild version `16.11.0.36601` still works with the `Qt5.12.11_x64_static_vs2019_160900` binaries.

One potential pitfall will be the `vcpkg` caching. We had [some issues](https://github.com/bitcoin/bitcoin/issues?q=is%3Aissue+vcpkg) with that in the past, namely whan a PR added or changed a dependency and the vcpkg couldn't deal with it. Since then the `vcpkg` manifest and binary caching features have been added. They may solve this problem.

fanquake commented:
ACK after rebase

hebasto commented:
Rebased ca9b9b9a2c0dd7ab767788be6512cfc34f1f6ee4 -> da8810e4f69b43e96ddf69400bec503451a0e847 ([pr21551.06](https://github.com/hebasto/bitcoin/commits/pr21551.06) -> [pr21551.07](https://github.com/hebasto/bitcoin/commits/pr21551.07)) due to the conflict with #22861.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -89,6 +73,64 @@ task:
+task:
+  name: "Win64 [unit tests, no functional tests] [msvc]"
+  windows_container:
+    cpu: 8
```
why 8 when the average usage is 2?

hebasto replied:
With `cpu: 4`:
 
![Screenshot from 2021-09-07 06-45-03](https://user-images.githubusercontent.com/32963518/132281199-c3c4f8d3-dd99-4b50-bab7-60d58a99af49.png)

And the total time with a populated `vcpkg` cache:

![Screenshot from 2021-09-07 06-46-09](https://user-images.githubusercontent.com/32963518/132281269-f262a40b-a91a-471a-8a5e-de59e59cf590.png)

MarcoFalke commented on .cirrus.yml:
```diff
@@ -89,6 +73,64 @@ task:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
 
+task:
+  name: "Win64 [unit tests, no functional tests] [msvc]"
```
```suggestion
  name: "Win64 native [unit tests, no functional tests] [msvc]"
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21551#issuecomment-913962298).

MarcoFalke commented on .cirrus.yml:
```diff
@@ -89,6 +73,64 @@ task:
+    cpu: 8
+    memory: 7424Mi
+    image: cirrusci/windowsservercore:visualstudio2019
+  << : *CREDITS_TEMPLATE
```
This looks expensive, especially with 8 CPUs and without `BASE_TEMPLATE` to skip

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/21551#issuecomment-913962298).

maflcko reviewed (commented):
Missing BASE_TEMPLATE to merge the latest branch?

hebasto committed:
ci: Add Windows task to Cirrus CI

hebasto committed:
ci: Drop AppVeyor CI integration

hebasto commented:
Updated da8810e4f69b43e96ddf69400bec503451a0e847 -> 97292b19140db32c6d85d63b70382e7bf60a55c4 ([pr21551.07](https://github.com/hebasto/bitcoin/commits/pr21551.07) -> [pr21551.08](https://github.com/hebasto/bitcoin/commits/pr21551.08), [diff](https://github.com/hebasto/bitcoin/compare/pr21551.07..pr21551.08)):

- addressed @MarcoFalke's comments

MarcoFalke commented:
ack

hebasto commented:
@sipsorcery 

Another (final?) review? :smile:

sipsorcery commented:
re-ACK 97292b19140db32c6d85d63b70382e7bf60a55c4.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -89,6 +76,73 @@ task:
+    - git config --global user.email "ci@ci.ci"
+    - git config --global user.name "ci"
+    - git config core.filemode false
+    - git reset --hard
```
maybe add a comment why this is needed?

hebasto replied:
Sure, I'm on it.

hebasto replied:
Done in #22909.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -89,6 +76,73 @@ task:
+    - python build_msvc\msvc-autogen.py
+    - msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minimal -noLogo
+  unit_tests_script:
+    - src\test_bitcoin.exe
```
why remove the `-l test_suite`?

hebasto replied:
:man_shrugging: Going to re-add right now.

hebasto replied:
Done in #22909.
